### ISSUE TITLE
ENH:  Change the cannonical dtypes to reflect actual underlying detpy

### ISSUE THREAD

**Issue by bashtage** [numpy/numpy#27704]
### Proposed new feature or change:

If you use things like `np.double` this is an alias for `np.float64`, at least on every platform I've ever used NumPy on.  Meanwhile, the docs claim that [`np.float64` is an alias for `np.double`](https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.float64).

This creates issues when using automatic typing in Sphinx docs since Sphinx sees a type like `NDArray[np.double]` and resolves it as `NDArray[np.float64]`, but then cannot find `np.float64` since it does not appear in intersphinx. There are other types that have the same issue, 

A partial list that I have encountered:

* `np.cdouble` -> `np.complex128`
* All of the types of the form `np.intXX` and `np.uintXX` are reported as aliases of other C types, but really are the canonical type used on a platform.  This also is surprising since when working cross platform it is more natural to use `np.int64` rather than `np.intp`, which may differ. 

Ideally, these types would no longer be the aliases in the docs and would become stand-alone definitions that can be targetted in intersphinx.

**Comment by rgommers** [comment]
Agreed, those reference docs should be changed to match the canonical (sized) names in https://numpy.org/devdocs/user/basics.types.html#relationship-between-numpy-data-types-and-c-data-types

**Comment by ngoldbaum** [comment]
Under the hood the sized integer types *are* aliases to the underlying C types, so at least for ints the docs are describing what NumPy is doing under the hood.

That said, maybe in 2024 we can just assume stdint.h is available and define the sized integer dtypes in terms of that?

That said I agree it’s not helpful that these internal historical details ate exposed in the docs. It would be better to not worry about how the sized types are defined internally and just update all the docs to use the sized dtypes explicitly and move away from the C data types as much as possible.

See also a similar ask from a few years ago, https://github.com/numpy/numpy/issues/20662

**Comment by bashtage** [comment]
Thanks @ngoldbaum and @rgommers. I could see two ways forward.  First, simply document the `stdint` types directly as stand along objects and stop referring them to as aliases.  This seems like a pretty free lunch. It may also be a bit more honest if there are some platforms where the C type doesn't actually map to the same stdint type, e.g., `long` to `intp`.

As for stdint, I think it is part of c99, which is a requirement for NumPy IIRC.

**Comment by seberg** [comment]
Agreed, this issue should be seen as a purely documentation improvement first.

FWIW, IIRC we changed the display name in Python to use `int64` or `float64` so that for the Python user the `int64` and `float64` names are the canonical ones and most can hopefully get away with not understanding the C names.

Yes to C-side changes: that is plausible, I had a branch once where I did a chunk of it.  The idea was that either `Int64DType` has multiple (singleton) instances or that `LongLongDType` is a subclass.  That would then allow for `np.dtype("l").char == "l"` to round-trip, while allowing us (starting with ufuncs) to use bit-sized names in C.
One could be ambitious and remove the `np.dtype("l").char == "l"` guarantee, but I am not sure i would want to start there.

(I.e. do not remove the type code distinction, but do merge them on the class level.)

If we are excited about that, we should open a new issue about it (if there isn't one already).
